src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java

```diff
@@ -239,14 +240,15 @@ public GHRepository repositoryFor(GitHubProject project) throws IOException {
    */
   public LocalRepository localRepositoryFor(GitHubProject project) throws IOException {
     Objects.requireNonNull(project, "On no! Project is null!");
+    synchronized (LOCAL_REPOSITORIES_INFO_LOCK) {
```

Comment:
`LOCAL_REPOSITORIES_INFO_LOCK` is used to synchronize access to `LOCAL_REPOSITORIES_INFO`. When you use it to synchronize access to `LOCAL_REPOSITORIES` as well, it means that threads are going to wait for the same lock even if they would like to access different fields. I see that this method calls `loadLocalRepositoryFo()` that accesses `LOCAL_REPOSITORIES_INFO`. However, it happens only when the cache doesn't have an element. I think it would be better to use different locks for those fields. But please make sure that you don't introduce a deadlock.

Reply:
To avoid this deadlock and synchronization issues. I suggest to replace `LRUMap` and `Map` with `ConcurrentHashMap`. More information can be found [here](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html). We would still need `synchronized` blocks while loading the repo and cleanup tasks. But at least the`maps` `localRepositoriesInfo` and `localRepositories` would be thread-safe.

@artem-smotrakov @mibo what do you think?

Reply:
`LRUMap` is necessary to keep only elements that were recently used. Maybe you could use `Collections.synchronizedMap()` that is mentioned here

https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/map/LRUMap.html

Reply:
👍